packages/components/src/panel/types.ts

```diff
@@ -1,3 +1,9 @@
+/**
```

Comment:
Here's also a few improvements that we could do to the JSDoc descriptions in this file:

- we should keep each line within 80 characters, for improved readability in IDEs
- For better formatting, we should surround technical terms (like prop names, component names, or values) with backticks, so that they get printed using a monospace font
- Since we're in the `PanelBody` props, we can also refer to `PanelBody` as "the component"
- In general, we should avoid documenting what I like to call "implementation details". Anything that we document can be assumed by the consumers of a component, and therefore we won't really be able to change it. For example, the fact that the `onToggle` callback is called when clicking the title, or after the open state is changed. The only thing that a user should know is that `onToggle` is called when the `PanelBody` component is toggled from its open state to its closed state, or viceversa

<details>

<summary>Here's the result of the tweaks that I applied on my machine</summary>

```diff
diff --git a/packages/components/src/panel/body.tsx b/packages/components/src/panel/body.tsx
index 329be31f37..c7ec030ce0 100644
--- a/packages/components/src/panel/body.tsx
+++ b/packages/components/src/panel/body.tsx
@@ -57,7 +57,7 @@ export function UnforwardedPanelBody(
 	};
 
 	// Ref is used so that the effect does not re-run upon scrollAfterOpen changing value.
-	const scrollAfterOpenRef = useRef< boolean | undefined >( undefined );
+	const scrollAfterOpenRef = useRef< boolean | undefined >();
 	scrollAfterOpenRef.current = scrollAfterOpen;
 	// Runs after initial render.
 	useUpdateEffect( () => {
@@ -101,12 +101,7 @@ export function UnforwardedPanelBody(
 
 const PanelBodyTitle = forwardRef(
 	(
-		{
-			isOpened,
-			icon,
-			title,
-			...props
-		}: WordPressComponentProps< PanelBodyTitleProps, 'button' >,
+		{ isOpened, icon, title, ...props }: PanelBodyTitleProps,
 		ref: React.ForwardedRef< any >
 	) => {
 		if ( ! title ) return null;
diff --git a/packages/components/src/panel/types.ts b/packages/components/src/panel/types.ts
index ef2fc65df1..37ddc32f42 100644
--- a/packages/components/src/panel/types.ts
+++ b/packages/components/src/panel/types.ts
@@ -43,6 +43,11 @@ export type PanelRowProps = {
 	children: React.ReactNode;
 };
 
+type ChildrenAsAFunctionProps = { opened: boolean };
+type ChildrenAsAFunction = (
+	props: ChildrenAsAFunctionProps
+) => React.ReactNode;
+
 export type PanelBodyProps = {
 	/**
 	 * Props that are passed to the `Button` component in title within the
@@ -59,16 +64,14 @@ export type PanelBodyProps = {
 	 * The content to display in the `PanelBody`.If a function is provided for
 	 * this prop, it will receive an object with the `opened` prop as an argument.
 	 */
-	children?:
-		| React.ReactNode
-		| ( ( props: { opened: boolean } ) => React.ReactNode );
+	children?: React.ReactNode | ChildrenAsAFunction;
 
 	/**
 	 * The CSS class to apply to the wrapper element.
 	 */
 	className?: string;
 	/**
-	 * An icon to be shown next to the PanelBody title.
+	 * An icon to be shown next to the title.
 	 */
 	icon?: JSX.Element;
 	/**
@@ -76,36 +79,41 @@ export type PanelBodyProps = {
 	 */
 	initialOpen?: boolean;
 	/**
-	 * A function that is called when the user clicks on the PanelBody title after the open state is changed.
+	 * A function that is called any time the component is toggled from its closed
+	 * state to its opened state, or viceversa.
 	 *
 	 * @default noop
 	 */
 	onToggle?: ( next: boolean ) => void;
 	/**
-	 * If opened is true then the Panel will remain open regardless of the initialOpen prop and the panel will be prevented from being closed.
+	 * When set to `true`, the component will remain open regardless of the
+	 * `initialOpen` prop and the panel will be prevented from being closed.
 	 */
 	opened?: boolean;
 	/**
-	 * Title of the PanelBody. This shows even when it is closed.
+	 * Title text. It shows even when it is closed.
 	 */
 	title?: string;
 	/**
-	 * Scrolls the content into view when visible.
-	 * This improves the UX when there are multiple stacking Panel Body components.
-	 * components in a scrollable container.
+	 * Scrolls the content into view when visible. This improves the UX when
+	 * multiple `PanelBody` are stacked in a scrollable container.
 	 *
 	 * @default true
 	 */
 	scrollAfterOpen?: boolean;
 };
 
-export type PanelBodyTitleProps = Omit< ButtonAsButtonProps, 'icon' > & {
+export type PanelBodyTitleProps = WordPressComponentProps<
+	Omit< ButtonAsButtonProps, 'icon' >,
+	'button',
+	false
+> & {
 	/**
-	 * An icon to be shown next to the PanelBody title.
+	 * An icon to be shown next to the title.
 	 */
 	icon?: JSX.Element;
 	/**
-	 * Whether or not the panel body is currently opened or not.
+	 * Whether or not the `PanelBody` is currently opened or not.
 	 */
 	isOpened?: boolean;
 	/**
```

</details>

Reply:
Thanks for the context! I had copied the text from the README initially and the following comment from the component since `scrollAfterOpen` isn't in the current README: 

https://github.com/WordPress/gutenberg/blob/8f23e951a0e6a4ddbb022dca3dac2dece948ae38/packages/components/src/panel/body.js#L63-L65

Should the README descriptions be identical to the JSDocs? And are there any other changes we make to the README? I'd like to at least match them to the order in the component file.

Reply:
> I had copied the text from the README initially and the following comment from the component since scrollAfterOpen isn't in the current README:

Thank you for the context!

The components package has been around for a while, and it didn't get much attention to such things until recently. Working on these migrations is also a way we can improve the docs, following the latest guidelines that we set for ourselves.

> Should the README descriptions be identical to the JSDocs? And are there any other changes we make to the README? I'd like to at least match them to the order in the component file.

Yup! That was actually going to be the focus of my next round of review.

Now that we're happy with the JSDoc descriptions in the `types.ts` file, we should apply those same descriptions to the README.

The goal is to have README, `types.ts` and the component's implemenations "synchronised":

- all props listed in `types.ts` should be listed (apart from deprecated/unstable ones)
- the types of the props should be the same across README and `types`
- the default values for props should be the same across code implementation, README and `types`
- the example snippet in the README should be applied to the default export (for `Panel`, [it already is](https://github.com/WordPress/gutenberg/blob/0e9295905a3a8ab4b6791fd90a9473be016b7b62/packages/components/src/panel/index.tsx#L30-L46))

Reply:
Thanks! And from looking at other examples with multiple types exported, it looks like I won't be adding info for `PanelBodyTitleProps` to the README. Is that right?

Reply:
> And from looking at other examples with multiple types exported, it looks like I won't be adding info for PanelBodyTitleProps to the README. Is that right?

Correct! We document in the README only the props of the components that are going to be used by consumers of the package (ie. `PanelBody`). `PanelTitle` is not an exported component, therefore no need to document it :)